### PR TITLE
PERF: concat_compat

### PR DESCRIPTION
**Pull request by jbrockmendel**
Fastpath for all-matching-dtype case.

### CODE REVIEW

File: pandas/core/dtypes/concat.py [hunk 1]
```diff
@@ -48,6 +57,24 @@ def concat_compat(to_concat, axis: AxisInt = 0, ea_compat_axis: bool = False):
...
+            return np.concatenate(to_concat_arrs, axis=axis)
+
+        to_concat_eas = cast("Sequence[ExtensionArray]", to_concat)
+        if ea_compat_axis:
+            # We have 1D objects, that don't support axis keyword
+            return obj._concat_same_type(to_concat_eas)
+        elif axis == 0:
+            return obj._concat_same_type(to_concat_eas)
+        else:
+            # e.g. DatetimeArray
```
Comment (mroeschke) [inline pandas/core/dtypes/concat.py:74]: Can we get here with EAs with _concat_same_type without the axis keyword e.g. Sparse?

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @jbrockmendel

**Comment by mroeschke** [comment]
@jbrockmendel I think there might be typing failures here. Mind taking a look? https://github.com/pandas-dev/pandas/actions/runs/4578025660/jobs/8084134007

**Comment by jbrockmendel** [comment]
yep, will make a PR shortly

**Pull request merged**